### DISCUSSION TITLE
C8602 nullability warning after comparing not null member to null

### DISCUSSION THREAD

**Discussion by joshudson** [dotnet/csharplang#8399] [General] [answered]
Reduced sample code:

```
class Login {
	private static SortedDictionary<string, Login> logins = null!;

	public static bool TryGetLogin(string email, [NotNullWhen(true)] out Login? login)
	{
		if (logins is null) InitLogins();
		return logins.TryGetValue(email, out login);
		// CS8602 'logins' may be null here. 
	}

	/// InitLogins() omitted for brevity. It's statically verifiable that after calling it, logins can't be null.
}
```

So the gist of this code is really simple. The property starts off null, but after InitLogins() it won't be null, and nothing ever assigns null to it again.

There's reasons why I'm not just exposing the member as a property but even if I did it doesn't seem to help.

Rewriting this to ```logins ??= InitLogins();``` doesn't work due to thread race issues at the point of the assignment. The assignment needs to happen inside `InitLogins()`.

So basically, the nullability warning is getting in the way here.

Since it's a static class variable not a local variable I'm kind of expecting it to throw out what it can infer between statements and just use the class level declaration every time. In the alternate hypothesis, different code can still get a null into it without a ! or a warning anywhere due to thread race issues.

**Comment by jnm2** [comment, marked as answer]
You can do this by applying `[MemberNotNull(nameof(logins))]` to the `InitLogins` method. Then the compiler will not warn in your sample, and it will check that `logins` is indeed not null for each return point within `InitLogins`.

**Comment by joshudson** [comment]
I hope my title's clear enough so the next person who hits this finds this discussion.
